builder/remotecontext/git/gitutils.go

```diff
@@ -98,22 +97,45 @@ func getRefAndSubdir(fragment string) (ref string, subdir string) {
 
 func fetchArgs(remoteURL string, ref string) []string {
 	args := []string{"fetch", "--recurse-submodules=yes"}
-	shallow := true
 
-	if urlutil.IsURL(remoteURL) {
-		res, err := http.Head(fmt.Sprintf("%s/info/refs?service=git-upload-pack", remoteURL))
-		if err != nil || res.Header.Get("Content-Type") != "application/x-git-upload-pack-advertisement" {
-			shallow = false
-		}
-	}
-
-	if shallow {
+	if supportsShallowCheckout(remoteURL) {
 		args = append(args, "--depth", "1")
 	}
 
 	return append(args, "origin", ref)
 }
 
+func supportsShallowCheckout(remoteURL string) bool {
+	// Check if the URL is an HTTP url
+	if urlutil.IsURL(remoteURL) {
```

Comment:
I left it in this form because I'm afraid we might eventually have special logic for other transport types. I think this really should look like a `switch protocol` statement with a default, but that doesn't quite work here.